File: Week 12/rust/1.StackImplementation.rs
/*
 * WEEK 12 - RUST DSA
 * Topic: Stack Implementation + Classic Problems
 * File: 1.StackImplementation.rs
 *
 * CONCEPT:
 *   LIFO container. Vec<T> in Rust is a growable array; push and pop are
 *   amortized O(1) — exactly the stack interface.
 *
 * KEY POINTS:
 *   - ArrayStack<T>: capped wrapper around Vec<T>.
 *   - MinStack: parallel Vec<i32> tracking running minimums.
 *   - Classic problems: balanced parens, next-greater, postfix.
 *
 * ALGORITHM / APPROACH:
 *   Same algorithms as Java/Python/C++. Differences are syntactic.
 *
 * RUST-SPECIFIC NOTES:
 *   - Vec<T>::push / Vec<T>::pop are the canonical stack operations.
 *   - pop returns Option<T> (None on empty) — no exceptions.
 *   - We use generics (`<T: Copy>` to keep `peek` simple).
 *   - For balanced-paren we use a Vec<char>; .chars() yields Unicode scalar
 *     values (i.e. proper code points, not bytes).
 *   - Postfix uses i32 and Rust's normal integer division (truncates toward
 *     zero, like Java/C++).
 *
 * DRY RUN:
 *   Same examples as Java file (omitted for brevity but identical).
 *
 * COMPLEXITY:
 *   push/pop/peek O(1); next-greater & postfix O(n).
 */

use std::collections::VecDeque;

// ---------- Generic capacity-bounded stack ----------

pub struct ArrayStack<T: Copy> {
    cap: usize,
    data: Vec<T>,
}

impl<T: Copy> ArrayStack<T> {
    pub fn new(capacity: usize) -> Self { Self { cap: capacity, data: Vec::with_capacity(capacity) } }

    pub fn push(&mut self, x: T) -> Result<(), &'static str> {
        if self.data.len() == self.cap { return Err("Stack overflow"); }
        self.data.push(x);
        Ok(())
    }

    pub fn pop(&mut self) -> Result<T, &'static str> {
        self.data.pop().ok_or("Stack underflow")
    }

    pub fn peek(&self) -> Result<T, &'static str> {
        self.data.last().copied().ok_or("Stack is empty")
    }

    pub fn is_empty(&self) -> bool { self.data.is_empty() }
    pub fn len(&self) -> usize { self.data.len() }
}

// ---------- MinStack ----------

pub struct MinStack {
    s: Vec<i32>,
    m: Vec<i32>,  // running minimums
}

impl MinStack {
    pub fn new() -> Self { Self { s: Vec::new(), m: Vec::new() } }

    pub fn push(&mut self, x: i32) {
        self.s.push(x);
        let new_min = match self.m.last() {
            Some(&cur) => cur.min(x),
            None => x,
        };
        self.m.push(new_min);
    }

    pub fn pop(&mut self) -> Option<i32> { self.m.pop(); self.s.pop() }
    pub fn peek(&self) -> Option<i32> { self.s.last().copied() }
    pub fn get_min(&self) -> Option<i32> { self.m.last().copied() }
}

// ---------- Classic problems ----------

pub fn is_balanced(s: &str) -> bool {
    let mut st: Vec<char> = Vec::new();
    for c in s.chars() {
        match c {
            '(' | '[' | '{' => st.push(c),
            ')' | ']' | '}' => {
                let want = match c { ')' => '(', ']' => '[', _ => '{' };
                match st.pop() {
                    Some(top) if top == want => {}
                    _ => return false,
                }
            }
            _ => {}  // ignore other chars
        }
    }
    st.is_empty()
}

pub fn next_greater_element(arr: &[i32]) -> Vec<i32> {
    let n = arr.len();
    let mut result = vec![-1; n];
    let mut st: VecDeque<usize> = VecDeque::new();  // indices, strictly decreasing
    for i in 0..n {
        while let Some(&top) = st.back() {
            if arr[i] > arr[top] {
                result[top] = arr[i];
                st.pop_back();
            } else { break; }
        }
        st.push_back(i);
    }
    result
}

pub fn eval_postfix(expr: &str) -> i32 {
    let mut st: Vec<i32> = Vec::new();
    for tok in expr.split_whitespace() {
        match tok {
            "+" | "-" | "*" | "/" => {
                let b = st.pop().unwrap();
                let a = st.pop().unwrap();
                st.push(match tok {
                    "+" => a + b,
                    "-" => a - b,
                    "*" => a * b,
                    _   => a / b,  // truncate toward zero
                });
            }
            _ => st.push(tok.parse().unwrap()),
        }
    }
    st.pop().unwrap()
}

fn main() {
    println!("=== Array Stack ===");
    let mut as_: ArrayStack<i32> = ArrayStack::new(10);
    as_.push(1).unwrap(); as_.push(2).unwrap(); as_.push(3).unwrap();
    println!("peek: {:?}", as_.peek().unwrap());
    println!("pop:  {:?}", as_.pop().unwrap());
    println!("pop:  {:?}", as_.pop().unwrap());
    println!("size: {}", as_.len());

    println!("\n=== Min Stack ===");
    let mut ms = MinStack::new();
    for v in [5, 3, 7, 2] { ms.push(v); }
    println!("peek: {:?}, min: {:?}", ms.peek(), ms.get_min());
    ms.pop();
    println!("After pop -- peek: {:?}, min: {:?}", ms.peek(), ms.get_min());

    println!("\n=== Balanced Parentheses ===");
    for t in ["([]{})", "(])", "((()))", "{[}]", ""] {
        println!("{:?}: {}", t, is_balanced(t));
    }

    println!("\n=== Next Greater Element ===");
    let arr = [4, 5, 2, 10, 8];
    println!("Array: {:?}", arr);
    println!("NGE:   {:?}", next_greater_element(&arr));

    println!("\n=== Postfix Evaluation ===");
    println!("\"3 4 +\" = {}", eval_postfix("3 4 +"));
    println!("\"3 4 + 2 *\" = {}", eval_postfix("3 4 + 2 *"));
    println!("\"15 7 1 1 + - / 3 * 2 1 1 + + -\" = {}",
        eval_postfix("15 7 1 1 + - / 3 * 2 1 1 + + -"));
}

/*
 * NOTES (vs. Java):
 * - Java's ArrayDeque is the recommended Stack. In Rust we use Vec or
 *   VecDeque. Both are growable arrays; VecDeque is ring-buffered.
 * - Java throws on underflow; Rust returns Result/Option so the caller is
 *   forced to handle the failure case.
 * - Generics in Rust are monomorphized like C++ templates (no runtime
 *   erasure as in Java).
 */

File: Week 12/cpp/1.StackImplementation.cpp
/*
 * WEEK 12 - C++ DSA
 * Topic: Stack Implementation + Classic Problems
 * File: 1.StackImplementation.cpp
 *
 * CONCEPT:
 *   LIFO container. push/pop/top/empty all O(1). STL provides std::stack
 *   (adapter) and std::vector (which makes a great underlying container).
 *
 * KEY POINTS:
 *   - ArrayStack<T>: fixed-capacity templated stack (educational).
 *   - MinStack: track running mins in a parallel stack.
 *   - Classic problems: balanced parens, next greater, postfix eval.
 *
 * ALGORITHM / APPROACH:
 *   See Week 12 Java file — same algorithms.
 *
 * C++-SPECIFIC NOTES:
 *   - We use std::vector<int> as the implementation — it's contiguous and
 *     amortized O(1) push_back / pop_back.
 *   - std::stack<int> would be the standard library choice; we roll our own
 *     to mirror the Java exercise.
 *   - Throw std::runtime_error / std::out_of_range on misuse.
 *
 * DRY RUN:
 *   Same as Java/Python files (balanced "([]{})" -> true; NGE [4,5,2,10,8]
 *   -> [5,10,10,-1,-1]; postfix "3 4 + 2 *" -> 14).
 *
 * COMPLEXITY:
 *   push/pop/peek: O(1); next-greater & postfix overall O(n).
 */

#include <iostream>
#include <sstream>
#include <stdexcept>
#include <stack>
#include <string>
#include <vector>

// ---------- Templated array-based stack (educational) ----------

template <typename T>
class ArrayStack {
public:
    explicit ArrayStack(std::size_t capacity) : cap(capacity) { data.reserve(capacity); }
    void push(const T& x) {
        if (data.size() == cap) throw std::runtime_error("Stack overflow");
        data.push_back(x);
    }
    T pop() {
        if (data.empty()) throw std::runtime_error("Stack underflow");
        T x = data.back(); data.pop_back(); return x;
    }
    const T& peek() const {
        if (data.empty()) throw std::runtime_error("Stack is empty");
        return data.back();
    }
    bool empty() const { return data.empty(); }
    std::size_t size() const { return data.size(); }
private:
    std::size_t cap;
    std::vector<T> data;
};

// ---------- MinStack ----------

class MinStack {
public:
    void push(int x) {
        s.push(x);
        m.push(m.empty() ? x : std::min(x, m.top()));
    }
    int pop() {
        int x = s.top(); s.pop(); m.pop();
        return x;
    }
    int peek() const { return s.top(); }
    int getMin() const { return m.top(); }
private:
    std::stack<int> s, m;
};

// ---------- Classic problems ----------

bool isBalanced(const std::string& s) {
    std::stack<char> st;
    for (char c : s) {
        if (c == '(' || c == '[' || c == '{') {
            st.push(c);
        } else if (c == ')' || c == ']' || c == '}') {
            if (st.empty()) return false;
            char top = st.top(); st.pop();
            if (c == ')' && top != '(') return false;
            if (c == ']' && top != '[') return false;
            if (c == '}' && top != '{') return false;
        }
    }
    return st.empty();
}

std::vector<int> nextGreaterElement(const std::vector<int>& arr) {
    int n = static_cast<int>(arr.size());
    std::vector<int> result(n, -1);
    std::stack<int> st;  // indices
    for (int i = 0; i < n; ++i) {
        while (!st.empty() && arr[i] > arr[st.top()]) {
            result[st.top()] = arr[i];
            st.pop();
        }
        st.push(i);
    }
    return result;
}

int evalPostfix(const std::string& expr) {
    std::stack<int> st;
    std::istringstream iss(expr);
    std::string tok;
    while (iss >> tok) {
        if (tok == "+" || tok == "-" || tok == "*" || tok == "/") {
            int b = st.top(); st.pop();
            int a = st.top(); st.pop();
            if (tok == "+") st.push(a + b);
            else if (tok == "-") st.push(a - b);
            else if (tok == "*") st.push(a * b);
            else st.push(a / b);  // C++ integer division truncates toward zero
        } else {
            st.push(std::stoi(tok));
        }
    }
    return st.top();
}

template<typename T>
std::string vec_str(const std::vector<T>& v) {
    std::ostringstream os;
    os << "[";
    for (std::size_t i = 0; i < v.size(); ++i) { if (i) os << ", "; os << v[i]; }
    os << "]";
    return os.str();
}

int main() {
    // Array stack
    std::cout << "=== Array Stack ===\n";
    ArrayStack<int> as(10);
    as.push(1); as.push(2); as.push(3);
    std::cout << "peek: " << as.peek() << "\n";
    std::cout << "pop:  " << as.pop()  << "\n";
    std::cout << "pop:  " << as.pop()  << "\n";
    std::cout << "size: " << as.size() << "\n";

    // Min stack
    std::cout << "\n=== Min Stack ===\n";
    MinStack ms;
    ms.push(5); ms.push(3); ms.push(7); ms.push(2);
    std::cout << "peek: " << ms.peek() << ", min: " << ms.getMin() << "\n";
    ms.pop();
    std::cout << "After pop -- peek: " << ms.peek() << ", min: " << ms.getMin() << "\n";

    // Balanced parens
    std::cout << "\n=== Balanced Parentheses ===\n";
    for (auto& t : {std::string("([]{})"), std::string("(])"),
                    std::string("((()))"), std::string("{[}]"),
                    std::string("")}) {
        std::cout << '"' << t << "\": " << std::boolalpha << isBalanced(t) << "\n";
    }

    // Next Greater Element
    std::cout << "\n=== Next Greater Element ===\n";
    std::vector<int> arr = {4, 5, 2, 10, 8};
    std::cout << "Array: " << vec_str(arr) << "\n";
    std::cout << "NGE:   " << vec_str(nextGreaterElement(arr)) << "\n";

    // Postfix eval
    std::cout << "\n=== Postfix Evaluation ===\n";
    std::cout << "\"3 4 +\" = "        << evalPostfix("3 4 +") << "\n";
    std::cout << "\"3 4 + 2 *\" = "    << evalPostfix("3 4 + 2 *") << "\n";
    std::cout << "\"15 7 1 1 + - / 3 * 2 1 1 + + -\" = "
              << evalPostfix("15 7 1 1 + - / 3 * 2 1 1 + + -") << "\n";

    return 0;
}

/*
 * NOTES (vs. Java):
 * - Java's recommended stack is ArrayDeque; here it's std::stack adapter.
 * - Templates give us a single generic ArrayStack<T> — Java uses generics
 *   that are erased at runtime; C++ templates are monomorphized.
 * - Integer division: both Java and C++ truncate toward zero (different
 *   from Python's // which floors).
 */
